Zend/zend_gc.c

```diff
@@ -389,52 +423,45 @@ static void gc_scan_black(zend_refcounted *ref)
 			ht = get_gc(&tmp, &zv, &n);
 			end = zv + n;
 			if (EXPECTED(!ht)) {
-				if (!n) return;
+				if (!n) return stack;
 				while (!Z_REFCOUNTED_P(--end)) {
-					if (zv == end) return;
+					if (zv == end) return stack;
 				}
 			}
 			while (zv != end) {
 				if (Z_REFCOUNTED_P(zv)) {
 					ref = Z_COUNTED_P(zv);
 					GC_REFCOUNT(ref)++;
-					if (GC_REF_GET_COLOR(ref) != GC_BLACK) {
-						gc_scan_black(ref);
-					}
+					stack = gc_refcounted_stack_push(stack, ref);
```

Comment:
Any particular reason why the GC_BLACK checks were moved? With this implementation you're going to be performing an unnecessary stack push+pop for non-black refs.

Reply:
The important part in the commit https://github.com/drealecs/php-src/commit/98b44236d8b5d7f1f275a63f7c5e41fe09eb8263 was adding the check on the start of the function so I removed it from previous points.
Yes, it adds a push, pop and also an extra function call.
Being a high level developer, I guess I tend to optimize code for readability and maintainability, not necessary for performance :). I just have to change this mindset a bit for this now.